src/strategies/FluentStrategyMixins.ts

```diff
@@ -0,0 +1,55 @@
+import {Arbitrary, FluentPick} from '../arbitraries'
+import {FluentResult} from '../FluentCheck'
+import {FluentStrategy, FluentStrategyInterface} from './FluentStrategy'
+
+type MixinConstructor<T = {}> = new (...args: any[]) => T
+type MixinStrategy = MixinConstructor<FluentStrategy>
+
+export function Random<TBase extends MixinStrategy>(Base: TBase) {
+  return class extends Base implements FluentStrategyInterface {
+    hasInput<K extends string>(arbitraryName: K): boolean {
+      return this.arbitraries[arbitraryName] !== undefined &&
+        this.arbitraries[arbitraryName].pickNum < this.arbitraries[arbitraryName].collection.length
+    }
+
+    getInput<K extends string, A>(arbitraryName: K): FluentPick<A> {
+      return this.arbitraries[arbitraryName].collection[this.arbitraries[arbitraryName].pickNum++]
+    }
+
+    handleResult() {}
+  }
+}
+
+export function Shrinkable<TBase extends MixinStrategy>(Base: TBase) {
+  return class extends Base {
+    shrink<K extends string>(arbitraryName: K, partial: FluentResult | undefined, depth: number) {
+      const shrinkedArbitrary = this.arbitraries[arbitraryName].arbitrary.shrink(partial!.example[arbitraryName])
+      this.arbitraries[arbitraryName].collection = this.buildArbitraryCollection(shrinkedArbitrary,
+        this.configuration.shrinkSize)
+    }
+  }
+}
+
+export function Dedupable<TBase extends MixinStrategy>(Base: TBase) {
+  return class extends Base {
+    isDedupable() {
+      return true
+    }
+  }
+}
+
+export function Cacheable<TBase extends MixinStrategy>(Base: TBase) {
```

Comment:
Semantically speaking:

Not sure if the verb should be cacheable. A strategy is not cacheable. Neither are the values. We are caching the values as part of the strategy, which is different.

Reply:
I agree. `Cached` seems more appropriate